[ZEPPELIN-2582][DOCS] docs for interpreter binding modes

1ambda opened the pull request:
### What is this PR for?

Updated `interpreter_binding_mode.md` since users are sometimes confused what this mode means and there is already opened JIRA issue. This documentation will be helpful to Zeppelin users.

**disclaimer: content was copied from [here](https://medium.com/@leemoonsoo/apache-zeppelin-interpreter-mode-explained-bae0525d0555) with author's consent.**


### What type of PR is it?
[Documentation]

### Todos

DONE

### What is the Jira issue?

[ZEPPELIN-2582](https://issues.apache.org/jira/browse/ZEPPELIN-2582)

### How should this be tested?

1. setup rvm 2.1.0+ and install required bundles for `docs/`
2. bundle exec jekyll serve --watch
3. http://localhost:4000/usage/interpreter/interpreter_binding_mode.html

### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/4968473/27527441-ac200154-5a86-11e7-8cd8-692aa4fe98f5.png)

![image](https://user-images.githubusercontent.com/4968473/27527446-b3003b24-5a86-11e7-92bb-7b2ae7ae9aa5.png)

![image](https://user-images.githubusercontent.com/4968473/27527449-b802b55c-5a86-11e7-867b-bf86ddc1692a.png)

![image](https://user-images.githubusercontent.com/4968473/27527456-be74afa8-5a86-11e7-9509-861f40ee2175.png)

![image](https://user-images.githubusercontent.com/4968473/27527462-c467abea-5a86-11e7-8d75-0b94962a27b2.png)

![image](https://user-images.githubusercontent.com/4968473/27527464-c6c4ebfa-5a86-11e7-852c-25655ef519dd.png)


### Questions:
* Does the licenses files need update? - NO
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - This PR is about docs.

1ambda commented:
CI failed but irrelevant.

- https://travis-ci.org/1ambda/zeppelin/jobs/246983015

cacti77 commented:
Thanks for doing this. I looked at the screenshots above. The diagrams are very helpful!  Are you also going to add the table I created at https://issues.apache.org/jira/browse/ZEPPELIN-2582?focusedCommentId=16024796&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16024796 or do you think it doesn't add anything useful?

1ambda commented:
@cacti77 Yes, it would be a good summary. Let me add it.

1ambda commented:
I added your summary table.

![image](https://user-images.githubusercontent.com/4968473/27566654-9be8346e-5b21-11e7-8e5d-4a264c9e7bb0.png)

1ambda commented:
CI failed but irrelevant.

zjffdu commented:
It is better to mention in what dimension of these modes. (per user or per note).
And from the code perspective, there's only one `InterpreterGroup` for scoped mode, but multiple sessions in one InterpreterGroup.

1ambda commented:
@zjffdu Thanks. Let me add the description into the doc.

1ambda commented:
@zjffdu enhanced description according to your suggestion. could you check that?

![image](https://user-images.githubusercontent.com/4968473/27579533-d75f5a90-5b62-11e7-869c-266c2af62999.png)

![image](https://user-images.githubusercontent.com/4968473/27579536-db94d644-5b62-11e7-9adc-2c4e593091b3.png)

zjffdu commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,88 @@ limitations under the License.
+</div>
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each Note. 
```
>> Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each Note. 

This is only correct when set scoped per note. As in another comment, we should mention in which dimension the scope is. And what does it mean for Per User and Per Note. The same for isolated mode.

1ambda replied:
![image](https://user-images.githubusercontent.com/4968473/27580640-41b10a8a-5b66-11e7-8226-f6339c871b8e.png)

Here is the description for that. If we need to repeat again there, I will add. Do we need?

1ambda commented:
@zjffdu Thanks for the review.

Now it looks like

![image](https://user-images.githubusercontent.com/4968473/27581456-bd729506-5b68-11e7-9ccf-7ccf5545bf89.png)

![image](https://user-images.githubusercontent.com/4968473/27581463-c0ed8754-5b68-11e7-9502-619c27bceec5.png)

1ambda commented:
CI failed but irrelevant.

zjffdu commented:
>> So, each Note have their own dedicated session but still it’s possible to share objects between different Interpreter Groups while they’re in the same JVM process.

Is this correct ? AFAIK, only sparkcontext is shared in scoped mode. Variable defined in session 1 can not be used in session 2

>> From the code perspective, there is only one `InterpreterGroup` for the scoped mode, but multiple sessions in one `InterpreterGroup`.

Sorry, I think we can remove this before we find a more proper way to explain. This might confuse users too.

1ambda commented:
@zjffdu regarding that, I have question as well.

- In isolated mode / shared mode, can't we share variables via `z.put?` and `z.get`?

zjffdu commented:
We can, but this is via `ResourcePool`. We might need to say that we could share object via resource pool. Otherwise users might be confused.

1ambda commented:
@zjffdu yes, it would be more clear. 

@jongyoul. As zjffdu said, variables (e.g scala variables, ...) in the scope mode can't be shared directly. is it correct?

jongyoul commented:
Yes, correct. Except `Shared`, user cannot share anything directly.

1ambda commented:
@jongyoul Thanks! 

I fixed the incorrect description. Now it looks like

![image](https://user-images.githubusercontent.com/4968473/27618720-f5f11c46-5bf8-11e7-9fd7-c9cd978582ae.png)

cacti77 commented:
> from the code perspective, there's only one InterpreterGroup for scoped mode, but multiple sessions in one InterpreterGroup.

If this is true then surely the diagram for scoped mode should only have one InterpreterGroup in it, but with multiple (Spark?) sessions inside that, with several interpreters inside each session?

Also, this 'per note' and 'per user' does appear to add another dimension. So, to explain what's happening you'd need one table for shared/scoped/isolated for the 'per user' mode and another table again for shared/scoped/isolated for the 'per note' mode.

It would also be good to have a column or table explaining how data can be shared, or not, between notebooks in the 3 different modes. As a Zeppelin user I find all these different combinations really confusing. Maybe an architectural diagram of how the different parts of Zeppelin talk to each other would help here. Or maybe the product itself should be simplified to remove a lot of these combinations, especially if they're not used much?!

1ambda commented:
@cacti77 So, you want to add some description for **per user** as well?

1ambda commented:
1. Added the **sharing objects** column.
2. Enhanced the description for **per user**

![image](https://user-images.githubusercontent.com/4968473/27667762-fa7d18f4-5cb7-11e7-907f-80f57b791548.png)

3. Added images for UI

![image](https://user-images.githubusercontent.com/4968473/27667773-061de62a-5cb8-11e7-9ff5-9f165e9293ef.png)

zjffdu commented:
@cacti77  @1ambda  How about diagram like this ?

![image](https://user-images.githubusercontent.com/164491/27667765-9ca2a6ac-5caf-11e7-8dcf-781252d40c8d.png)
![image](https://user-images.githubusercontent.com/164491/27667735-7744299e-5caf-11e7-9fb7-471a0abe5638.png)

1ambda commented:
@zjffdu I like that diagram, it's more precise and accurate. But would be nice to

- colorize
- and attach note UI so that users easily understand 

like 

![image](https://user-images.githubusercontent.com/4968473/27667820-595e45fa-5cb8-11e7-8787-82752433d7dd.png)

1ambda commented:
@cacti77 @zjffdu Please tell me if this is incorrect

![image](https://user-images.githubusercontent.com/4968473/27667834-6b63f97a-5cb8-11e7-9c0b-3ddf2a28845c.png)

1ambda commented:
@zjffdu Let me work on the new diagram based on yours, and will share it here.

1ambda commented:
Now it looks like

![image](https://user-images.githubusercontent.com/4968473/27670467-180708fc-5cc8-11e7-8ffc-bdf4ca224e1b.png)

![image](https://user-images.githubusercontent.com/4968473/27670469-1ba75f8e-5cc8-11e7-99bc-f3d3c0deef97.png)

![image](https://user-images.githubusercontent.com/4968473/27670473-1f54b30c-5cc8-11e7-9543-4b31beeed77c.png)

![image](https://user-images.githubusercontent.com/4968473/27670479-282b4de2-5cc8-11e7-9ca5-df3bd688fa25.png)

![image](https://user-images.githubusercontent.com/4968473/27670482-2b86f04a-5cc8-11e7-8623-74b1831fb542.png)

![image](https://user-images.githubusercontent.com/4968473/27670486-2fa69400-5cc8-11e7-8b9b-27033fcd22db.png)

felixcheung commented:
awesome!

felixcheung commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+Interpreter is a JVM process that communicates to Zeppelin daemon using thrift. 
+Each Interpreter process can have Interpreter Groups, and each interpreter instance belongs to this Interpreter Group.
```
does this mean to be one group?
`Each Interpreter process can have an Interpreter Group`
?
or
`Each Interpreter process can have multiple Interpreter Groups`
for clarity?

felixcheung commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+</div>
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each note. (in case of **per note**) 
```
`multiple Interpreter Group` -> `multiple Interpreter Groups`

felixcheung commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+Each Interpreter implementation may have different characteristics depending on the back end system that they integrate. And 3 interpreter modes can be used differently.
+Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mote, as an example. 
```
`mote` -> `note`

1ambda commented:
Many sentences should be updated since our major unit is the session, not interpreter group in the screenshots. Let me updated.

zjffdu commented:
Thanks @1ambda for the help on the updated diagram, it looks create. Just leave a minor comment.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+(See [here](../../development/writing_zeppelin_interpreter.html) to understand more about its internal structure.) 
+
+Zeppelin provides 3 different modes to run interpreter process: **shared**, **scoped** and **isolated**.   
+Also, user can specify the scope of this mode as well: **per user** or **per note**.  
```
The screenshot shows 3 scopes: Globally, Per Note, Per User, so shouldn't Globally be mentioned too? Also, in the UI itself under Option, does it really mean "The interpreter will be instantiated...", or should it be "The interpreter **group** will be instantiated..."?

1ambda replied:
1. Yeah. `Globally` means the `shared` mode. 
2. **The interpreter group will be instantiated** more makes sense.

cacti77 commented:
Sorry, I've never done a PR before so I'm not familiar with how best to contribute my suggestions. I'm also finding it difficult to see what the final docs page would look like as a whole.

1. The diagrams are looking great but shouldn't we also show the InterpreterGroup since it is mentioned in the UI itself and in the text of this help page? Does the InterpreterGroup contain the session, or is it vice versa?

1. In the first set of 3 diagrams maybe distinguish the Interpreters in each session (e.g., Interpreter 1, Interpreter 2, Interpreter 3; don't use A, B, C because you've already used that to distinguish the notes!).

1. Thanks for adding the extra table column: typo (occurs twice): share objets --> share obje**c**ts (and remove ')' after "via ResourcePool")

1ambda commented:
@cacti77 

Your comment is helpful. You can preview the page like

- [setup rvm 2.1.0+ and install required bundles for docs/](https://github.com/apache/zeppelin/tree/master/docs#build-documentation)
- bundle exec jekyll serve --watch
- open `http://localhost:4000/usage/interpreter/interpreter_binding_mode.html`

1. `1 InterpreterGroup` matchs `1 Interpreter Process`, it might contains multiple sessions (e.g **scope mode**
2. I will update it. 
3. I will udpate it.

cacti77 commented:
@1ambda Thanks for the tip. I was hoping there'd be an easier way to see the web page; oh well:).

> 1 InterpreterGroup matchs 1 Interpreter Process

Are you sure about this?@felixcheung above said:

> Each Interpreter process can have multiple Interpreter Groups

I still think it would be good to distinguish Interpreter Groups from the Interpreter Process in the diagrams. After looking at the code for [InterpreterGroup ](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterGroup.java) it makes a bit more sense to me now.  One InterpreterGroup can be associated with multiple sessions, and each session id is associated with a list of interpreters. I'm just not sure yet [how the sessions are generated with respect to notes, users, or their combination](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterGroup.java#L41-L44)

1ambda commented:
@cacti77 

1. ![image](https://user-images.githubusercontent.com/4968473/27730825-a1adf6a8-5dc5-11e7-9ff2-425e23fe4426.png)

2. Could you provide an example diagram? you can duplicate and edit [this](https://docs.google.com/presentation/d/1Y8lAJFlMO-Jm6ObqcTPmZa9_S9NfyFDsv93xSXvO7l0/edit?usp=sharing) with the description to be included?

cacti77 commented:
@1ambda Thanks; I can try - I've requested edit permission on that doc.

cacti77 commented:
@1ambda Never mind; I'll edit my own copy as you suggested previously.

cacti77 commented:
@1ambda Look [here ](https://docs.google.com/presentation/d/1eIdnReyFxTAAOKTwc_a_mynhlt9mNq6GQAAewFP00PI/edit?usp=sharing) please. I've added 1, 2 or 3 to the Interpreters in the first 3 diagrams. I've only got the patience to add the InterpreterGroup to the very first diagram, but it should be added to all 6 diagrams. In the two scoped-mode diagrams (diagrams 2 and 5) the InterpreterGroup should contain all 3 sessions. In the last 3 diagrams I presume the red SparkContext box should be _outside_ the InterpreterGroup boxes since the code for the latter doesn't appear to directly reference the SparkContext. But @zjffdu should be able to confirm this is true.

Hope this helps you a bit!

zjffdu commented:
Thanks @cacti77 & @1ambda. The diagram is correct. One interpreter process would have only one InterperterGroup, and InterpreterGroup could have one session for shared/isolated mode, and multiple sessions for scoped mode.

1ambda commented:
Thanks for the review.

@felixcheung Resolved all comments.
@zjffdu @cacti77 Update imgs and description. See the last 2 commits.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
```
Can't share directly, but it's possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+**isolated** | Has its own Interpreter Process | One notebook not affected directly by other notebooks (**per note**) | Can't share data between notebooks easily (**per note**) | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
```
Can't share directly, but it's possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+Interpreter is a JVM process that communicates with Zeppelin daemon using thrift. 
+Each Interpreter process can have a interpreter group, and each interpreter instance belongs to this interpreter group.
```
Each interpreter process has a single interpreter group, and this interpreter group can have one or more instances of an interpreter.

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+(See [here](../../development/writing_zeppelin_interpreter.html) to understand more about its internal structure.) 
+
+Zeppelin provides 3 different modes to run interpreter process: **shared**, **scoped** and **isolated**.   
+Also, user can specify the scope of these mode as well: **per user** or **per note**.  
```
Also, the user can specify the scope of these modes as well: **per user** or **per note**.

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+Also, user can specify the scope of these mode as well: **per user** or **per note**.  
+These 3 modes give flexibility to fit Zeppelin into any type of use cases.
+
+In this documentation, we mainly discuss the combination of **per note** mode with **shared**, **scoped** and **isolated** modes for explanation. 
```
In this documentation, we mainly discuss the **per note** scope in combination with the **shared**, **scoped** and **isolated** modes.

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple sessions serve each note. (in case of **per note**) 
+So, each note have their own dedicated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
```
In Scoped mode, Zeppelin still runs a single interpreter JVM process but, in the case of **per note** scope, each note runs in its own dedicated session. (Note it is still possible to share objects between these notes via [ResourcePool](../../interpreter/spark.html#object-exchange).)

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+**Isolated** mode runs separate interpreter process for each note. (in case of **per note**) 
+So, each note have absolutely isolated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
```
**Isolated** mode runs a separate interpreter process for each note in the case of **per note** scope.
So, each note has an absolutely isolated session. (But it is still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange).)

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
```
+**shared** | Shares a single session in a single interpreter process (JVM) | Low resource utilization and it's easy to share data between notebooks | All notebooks are affected if the interpreter process dies | Can share directly

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+<br/>
+
+Each Interpreter implementation may have different characteristics depending on the back end system that they integrate. And 3 interpreter modes can be used differently.
+Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mdoe, as an example. 
```
Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mode, as an example.

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+
+In Scoped mode, each note has its own Scala REPL. 
+So variable defined in a note can not be read or overridden in another note. 
+However, still single SparkContext serves all the sessions. 
```
However, a single SparkContext still serves all the sessions.

1ambda replied:
Fixed.

cacti77 commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+In Scoped mode, each note has its own Scala REPL. 
+So variable defined in a note can not be read or overridden in another note. 
+However, still single SparkContext serves all the sessions. 
+And all the jobs are submitted to this SparkContext and fair scheduler schedules the job. 
```
And all the jobs are submitted to this SparkContext and the fair scheduler schedules the jobs.

1ambda replied:
Fixed.

cacti77 commented:
@1ambda Thanks for finishing the diagrams. I would probably put the SparkContext inside the interpreter process but outside the InterpreterGroup; but if @zjffdu is happy with the diagrams I guess it doesn't matter. I've gone through interpreter_binding_mode.md, tidying up the English some more. I'm done with my edits now, thanks.

cacti77 commented:
Apologies @1ambda for not interacting with this PR properly; I may have overlooked some of your recent edits in the version of interpreter_binding_mode.md I commented on yesterday.

1ambda commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+**isolated** | Has its own Interpreter Process | One notebook not affected directly by other notebooks (**per note**) | Can't share data between notebooks easily (**per note**) | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+
+In case of **per user** (available on multi-user environment), Zeppelin manages interpreter sessions per user. For example,
```
Fixed.

1ambda commented:
resolved @cacti77's all comments

zjffdu commented on docs/usage/interpreter/interpreter_binding_mode.md:
```diff
@@ -25,4 +25,97 @@ limitations under the License.
 
+<br/>
+
+Interpreter is a JVM process that communicates with Zeppelin daemon using thrift. 
```
Interpreter -> Interpreter Process

1ambda replied:
fixed.

1ambda commented:
resolved @zjffdu's comment

1ambda commented:
Please let me know if we need to add / improve / modify something.

zjffdu commented:
LGTM

cacti77 commented:
@1ambda And it looks good to me too; thanks for being patient and making all those changes!

1ambda commented:
it was an honor to make more correct, concise documentation with your help. 

Merge if no more discussion.

cacti77 commented:
@1ambda One question please: Do these docs constitute a new web page in Zeppelin's docs? I'm just wondering how a user would navigate to this new information on binding modes from the Docs home page. Will the new page be linked to from under http://zeppelin.apache.org/docs/0.7.2/manual/interpreters.html#interpreter-binding-mode for example? Or will it become part of the interpreters.html page itself?

1ambda commented:
@cacti77 it's provided as of 0.8.0-SNAPSHOT. 

- http://zeppelin.apache.org/docs/0.8.0-SNAPSHOT/usage/interpreter/interpreter_binding_mode.html

here are the background

- we re-organized contents structure of docs in 0.8.0-SNAPSHOT
- the interpreter binding page was created at that time (so it's currently empty)

cacti77 commented:
Thanks @1ambda . So, looking at #2371, it looks like a user would simply click on the dropdown and then Interpreter > Interpreter Binding Mode to access the new web page. Is that correct? Basically, as long as all the info about interpreter binding modes is in one place and easy to find, that's all I care about:).

1ambda commented:
Sure, users can easily access the doc.

1ambda commented:
Deployed 

- http://zeppelin.apache.org/docs/0.8.0-SNAPSHOT/usage/interpreter/interpreter_binding_mode.html
